### PR TITLE
fix(refs dplan-11402, dplan-11403):  Refactor logic to display Permissions for selected Phase

### PR DESCRIPTION
**Pull request by salisdemos**
https://demoseurope.youtrack.cloud/issue/DPLAN-11402/Hinweis-Grundeinstellungen-Angaben-der-Rechte-zu-Verfahrensschritten-fehlen

https://demoseurope.youtrack.cloud/issue/DPLAN-11403/Layout-in-Grundeinstellungen-nicht-einheitlich

To fix the Bug, I hat to move the logic from twig and wrapper to a dedicated Component.

### How to review/test

Goto Grundeinstellungen and change the phases. See if the Permission text is displayed and makes sense.

### CODE REVIEW

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+            tooltip: helpText
+          }"
+          :name="fieldName"
+          :options="phaseOptions"
+          required
+          v-model="selectedPhase" />
+      </div>
+
+      <div
+        v-if="true || hasPermission('field_phase_iterator')"
```
Comment (hwiem) [inline client/js/components/procedure/basicSettings/ParticipationPhases.vue:17]: Is this correct?

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        class="w-4/12 inline-block relative">
+        <dp-label
```
Comment (hwiem) [inline client/js/components/procedure/basicSettings/ParticipationPhases.vue:19]: Shouldn't there be an input with the id `r_public_participation_phase_iteration`?

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  },
+
+  props: {
+    autoswitchHint: {
+      type: String,
+      required: false,
+      default: ''
+    },
+
+    labelText: {
```
Comment (hwiem) [inline client/js/components/procedure/basicSettings/ParticipationPhases.vue:68]: nitpick: please order the props alphabetically

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      default: ''
+    },
+
+    initSelectedPhase: {
+      type: String,
+      required: false,
+      default: ''
+    },
+
+    iterator: {
```
Comment (hwiem) [inline client/js/components/procedure/basicSettings/ParticipationPhases.vue:92]: Why did you call this prop `iterator`? (I can't really imagine what this is, but I see that iterator is used in other places, too. Maybe a doc block would help)
I think we should always add a validator or a doc block to explain the properties an object prop should have, so it's easy to see which data we need to pass into a component

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    permissionMessageText () {
+      const currentPhase = this.phaseOptions.find(option => option.value === this.selectedPhase)
+      /*
+       * Generated Trans-Keyes:
+       *
+       * 'permissionset.hidden'
+       * 'permissionset.read'
+       * 'permissionset.write'
+       */
+      return this.permissionMessage + Translator.trans(`permissionset.${currentPhase.permissionset}`)
```
Comment (hwiem) [inline client/js/components/procedure/basicSettings/ParticipationPhases.vue:133]: nitpick: this way it's clearer what the doc block refers to

```suggestion
      const currentPhase = this.phaseOptions.find(option => option.value === this.selectedPhase)
       /*
       * Generated Trans-Keyes:
       *
       * 'permissionset.hidden'
       * 'permissionset.read'
       * 'permissionset.write'
       */
      const permissionsetMessage =  Translator.trans(`permissionset.${currentPhase.permissionset}`)
     
      return `${this.permissionMessage} ${permissionsetMessage}`
```

File: templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig [hunk 6]
```diff
@@ -488,7 +451,10 @@
 
                                     <dp-date-range-picker
                                         class="u-2-of-4"
-                                        data-cy="procedureSettingsDate"
+                                        :data-cy="{
```
Comment (hwiem) [inline templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:454]: Why this change? Doesn't `dataCy` have to be a string?

File: templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig [hunk 7]
```diff
@@ -564,64 +530,29 @@
...
+                                            label: phase.name,
+                                            permissionset: phase.permissionset,
+                                        }]) %}
+                                    {% endfor %}
+                                    <participation-phases
+                                        autoswitch-hint="{{ 'period.autoswitch.hint'|trans({ phase: 'procedure.phases.external.evaluating'|trans }) }}"
+                                        field-name="r_publicParticipationPhase"
+                                        help-text="{{ 'text.procedure.edit.external.change_phase'|trans }}"
+                                        label-text="{{ "procedure.public.phase"|trans }}"
+                                        :participation-phases="['participation', 'earlyparticipation', 'anotherparticipation']"
```
Comment (hwiem) [inline templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:546]: Is it correct to hard code the phases here?

File: templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig [hunk 8]
```diff
@@ -642,7 +573,10 @@
                                     } %}
                                     <dp-date-range-picker
                                         class="u-2-of-4"
-                                        data-cy="procedureSettingsParticipationDate"
+                                        :data-cy="{
```
Comment (hwiem) [inline templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:576]: Shouldn't this prop be a string?

**Review by hwiem — CHANGES_REQUESTED** [review]

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue (hunk 1 above)
Comment (salisdemos) [reply client/js/components/procedure/basicSettings/ParticipationPhases.vue:17]: obviousely not.
```suggestion
        v-if="hasPermission('field_phase_iterator')"
```

File: templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig (hunk 7 above)
Comment (salisdemos) [reply templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:547]: At least as correct as it was before.
See removed line templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:112

File: templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig (hunk 8 above)
Comment (salisdemos) [reply templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:578]: Means. It has to be changed with the next demosplan-ui release. But for now it has to be an object. The core-PR that changed this got merged before the new demosplan-ui was released

File: templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig (hunk 6 above)
Comment (salisdemos) [reply templates/bundles/DemosPlanCoreBundle/DemosPlanProcedure/administration_edit.html.twig:455]: see below

File: client/js/components/procedure/basicSettings/ParticipationPhases.vue (hunk 4 above)
Comment (salisdemos) [reply client/js/components/procedure/basicSettings/ParticipationPhases.vue:87]: Now as You question: I could "hardcode" everything but the iterator.value. Would You prefer that?
Otherwise, I would follow Your suggestion and add a doc-Block and a validator.

Comment (hwiem) [reply client/js/components/procedure/basicSettings/ParticipationPhases.vue:87]: If it is unlikely that we want to pass in other values for these properties, I think we should "hard code" them. But a short explanation for the `iterator.value` might still be useful. :)

Comment (salisdemos) [reply client/js/components/procedure/basicSettings/ParticipationPhases.vue:91]: I missed, that we still need the unique name. Here we go 9808008566ea3c69576b2dd86890ce78d9b8714b

File: tests/frontend/unit/ParticipationPhases.spec.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }, {
+      label: 'My Label B',
+      permissionset: 'hidden',
+      value: 'fifth'
+    }],
+    initSelectedPhase: 'first',
+    iterator: {
+      label: 'Iterator Label',
+      name: 'iterator_input_name',
+      tooltip: 'Tooltip text for the itereator',
```
Comment (hwiem) [inline tests/frontend/unit/ParticipationPhases.spec.js:30]:
```suggestion
      tooltip: 'Tooltip text for the iterator',
```

File: tests/frontend/unit/ParticipationPhases.spec.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+  it('should be an object', () => {
+    expect(typeof ParticipationPhases).toBe('object')
+  })
+
+  it('should be named ParticipationPhases', () => {
+    expect(ParticipationPhases.name).toBe('ParticipationPhases')
+  })
+
+  it('should set the "inParticipation"-State as expected', async () => {
```
Comment (hwiem) [inline tests/frontend/unit/ParticipationPhases.spec.js:44]: I think tests should also document how a component is supposed to work, so I find it useful to say it explicitly (rather than 'as expected'. what is expected?), e.g. 'sets `inParticipation` to true if the selected Phase is a participation phase'. I would also test only one case in a single test, so I'd add separate tests for when `inParticipation` should be `false`.

File: tests/frontend/unit/ParticipationPhases.spec.js [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    expect(wrapper.vm.isInParticipation).toBe(false)
+
+    await wrapper.setData({
+      selectedPhase: 'noneAtall',
+    })
+
+    expect(wrapper.vm.isInParticipation).toBe(false)
+  })
+
+  it('should set the Permission-Message-Text as expected', async () => {
```
Comment (hwiem) [inline tests/frontend/unit/ParticipationPhases.spec.js:67]: see above

File: tests/frontend/unit/ParticipationPhases.spec.js (hunk 10 above)
Comment (salisdemos) [inline tests/frontend/unit/ParticipationPhases.spec.js:44]:
```suggestion
  it('should set the "inParticipation"-State to true if the selected Phase occures in the participationPhases Array', async () => {
```

File: tests/frontend/unit/ParticipationPhases.spec.js (hunk 11 above)
Comment (salisdemos) [reply tests/frontend/unit/ParticipationPhases.spec.js:67]:
```suggestion
  it('should set the Permission-Message-Text depending on the selected participation phase permissionset ', async () => {
```

File: tests/frontend/unit/ParticipationPhases.spec.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+  it('should be an object', () => {
+    expect(typeof ParticipationPhases).toBe('object')
+  })
+
+  it('should be named ParticipationPhases', () => {
+    expect(ParticipationPhases.name).toBe('ParticipationPhases')
+  })
+
+  it('should set the "inParticipation"-State to true if the selected Phase occures in the participationPhases Array', async () => {
```
[the remark this answers is not in the document]
Comment (salisdemos) [reply tests/frontend/unit/ParticipationPhases.spec.js:44]: Maybe you have an Idea for a fitting description? I am quite new to this field so have to get used to write proper descriptions still

[the remark this answers is not in the document]
Comment (hwiem) [reply tests/frontend/unit/ParticipationPhases.spec.js:44]: I don't really have a good idea, but how about

`it('sets the 'inParticipation' state correctly depending on whether the selected Phase is a participation phase',`

? I'd like to keep implementation details out of the description as much as possible.

[the remark this answers is not in the document]
Comment (salisdemos) [reply tests/frontend/unit/ParticipationPhases.spec.js:44]:
```suggestion
  it('sets the 'inParticipation' state correctly depending on whether the selected Phase is a participation phase', async () => {
```

**Review by hwiem — APPROVED** [review]
lgtm, code review only
